components/tidb_query/src/codec/convert.rs

```diff
@@ -1692,4 +1828,129 @@ mod tests {
+
+        for (s, char_num, cs) in cases {
+            ft.set_charset(cs.to_string());
+            ft.set_flen(char_num);
+            let bs = s.as_bytes();
+            let r = produce_str_with_specified_tp(&mut ctx, Cow::Borrowed(bs), &ft, false);
+            assert!(r.is_ok(), "{}, {}, {}", s, char_num, cs);
+            let p = r.unwrap();
+
+            if cs == charset::CHARSET_UTF8MB4 || cs == charset::CHARSET_UTF8 {
+                let ns: String = s.chars().take(char_num as usize).collect();
+                assert_eq!(p.as_ref(), ns.as_bytes(), "{}, {}, {}", s, char_num, cs);
+            } else {
+                assert_eq!(
+                    p.as_ref(),
+                    &bs[..(char_num as usize)],
+                    "{}, {}, {}",
+                    s,
+                    char_num,
+                    cs
+                );
+            }
+        }
+
+        let cases = vec![
+            // branch 3
+            ("世界，中国", 20, charset::CHARSET_ASCII),
+            ("世界，中国", 30, charset::CHARSET_ASCII),
+            ("世界，中国", 50, charset::CHARSET_ASCII),
+        ];
+
+        use tidb_query_datatype::FieldTypeAccessor;
+
+        let mut cfg = EvalConfig::new();
+        cfg.set_flag(Flag::TRUNCATE_AS_WARNING);
+        let mut ctx = EvalContext::new(Arc::new(cfg));
+        let mut ft = FieldType::new();
+
+        <FieldType as FieldTypeAccessor>::set_tp(&mut ft, FieldTypeTp::String);
+        <FieldType as FieldTypeAccessor>::set_collation(&mut ft, Collation::Binary);
+
+        for (s, char_num, cs) in cases {
+            ft.set_charset(cs.to_string());
+            ft.set_flen(char_num);
+            let bs = s.as_bytes();
+            let r = produce_str_with_specified_tp(&mut ctx, Cow::Borrowed(bs), &ft, true);
+            assert!(r.is_ok(), "{}, {}, {}", s, char_num, cs);
+
+            let p = r.unwrap();
+            assert_eq!(p.len(), char_num as usize, "{}, {}, {}", s, char_num, cs);
+        }
+    }
+
+    #[test]
+    fn test_produce_dec_with_specified_tp() {
+        use std::str::FromStr;
+
+        let cases = vec![
+            // branch 1
+            (
+                Decimal::from_str("11.1").unwrap(),
+                2,
+                2,
+                max_or_min_dec(false, 2u8, 2u8),
+            ),
+            (
+                Decimal::from_str("-111.1").unwrap(),
+                2,
+                2,
+                max_or_min_dec(true, 2u8, 2u8),
+            ),
+            // branch 2
+            (
+                Decimal::from_str("-1111.1").unwrap(),
+                5,
+                1,
+                Decimal::from_str("-1111.1").unwrap(),
+            ),
+            (
+                Decimal::from_str("-111.111").unwrap(),
+                5,
+                2,
+                Decimal::from_str("-111.11").unwrap(),
+            ),
+        ];
+
+        let mut cfg = EvalConfig::new();
+        cfg.set_flag(Flag::TRUNCATE_AS_WARNING);
```

Comment:
ditto

Reply:
here we have two set flag, so use from_flag is not ok

Reply:
It's OK to use two or more flag, e.g: `EvalConfig::from_flag(Flag1 | Flag2 | Flag3)`